File: Example/javaapp/src/main/java/com/stripe/example/javaapp/fragment/discovery/ReaderAdapter.java
package com.stripe.example.javaapp.fragment.discovery;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.stripe.example.javaapp.R;
import com.stripe.example.javaapp.viewmodel.DiscoveryViewModel;
import com.stripe.stripeterminal.external.models.Location;
import com.stripe.stripeterminal.external.models.Reader;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Our [RecyclerView.Adapter] implementation that allows us to update the list of readers
 */
public class ReaderAdapter extends RecyclerView.Adapter<ReaderHolder> {
    @NotNull private final DiscoveryViewModel viewModel;
    @NotNull private final LayoutInflater inflater;
    @NotNull private List<? extends Reader> readers;
    @Nullable private Location locationSelection = null;

    public ReaderAdapter(@NotNull DiscoveryViewModel viewModel, LayoutInflater inflater) {
        super();
        this.inflater = inflater;
        this.viewModel = viewModel;
        if (viewModel.readers.getValue() == null) {
            readers = new ArrayList<>();
        } else {
            readers = viewModel.readers.getValue();
        }
    }

    @BindingAdapter("readers")
    public static void bindReaders(
            @NotNull RecyclerView recyclerView,
            @NotNull List<? extends Reader> readers
    ) {
        if (recyclerView.getAdapter() instanceof ReaderAdapter) {
            ((ReaderAdapter) recyclerView.getAdapter()).updateReaders(readers);
        }
    }

    public void updateLocationSelection(@Nullable Location location) {
        this.locationSelection = location;
        notifyDataSetChanged();
    }

    public void updateReaders(List<? extends Reader> readers) {
        this.readers = readers;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return readers.size();
    }

    @Override
    public void onBindViewHolder(@NotNull ReaderHolder holder, int position) {
        holder.bind(readers.get(position), locationSelection);
    }

    @NotNull
    @Override
    public ReaderHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        assert viewModel.readerClickListener != null;
        return new ReaderHolder(inflater.inflate(R.layout.list_item_card, parent, false), viewModel.readerClickListener);
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/fragment/discovery/ReaderAdapter.kt
package com.stripe.example.fragment.discovery

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.BindingAdapter
import androidx.recyclerview.widget.RecyclerView
import com.stripe.example.R
import com.stripe.example.viewmodel.DiscoveryViewModel
import com.stripe.stripeterminal.external.models.Location
import com.stripe.stripeterminal.external.models.Reader

object ItemsBindingAdapter {
    @BindingAdapter("items")
    @JvmStatic
    fun RecyclerView.bindItems(items: List<Reader>) {
        val adapter = adapter as ReaderAdapter
        adapter.updateReaders(items)
    }
}

/**
 * Our [RecyclerView.Adapter] implementation that allows us to update the list of readers
 */
class ReaderAdapter(
    private val viewModel: DiscoveryViewModel,
    private val inflater: LayoutInflater,
) : RecyclerView.Adapter<ReaderHolder>() {
    private var readers: List<Reader> = viewModel.readers.value ?: listOf()
    private var locationSelection: Location? = null

    fun updateReaders(readers: List<Reader>) {
        this.readers = readers
        notifyDataSetChanged()
    }

    fun updateLocationSelection(location: Location?) {
        locationSelection = location
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return readers.size
    }

    override fun onBindViewHolder(holder: ReaderHolder, position: Int) {
        holder.bind(readers[position], locationSelection)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReaderHolder {
        return ReaderHolder(
            parent = inflater.inflate(R.layout.list_item_card, parent, false),
            clickListener = viewModel.readerClickListener!!,
        )
    }
}
